### PR TITLE
makefile.iotlab.single: add IoT-LAB testbed support for one node

### PR DESCRIPTION
**Pull request by cladmi**
Add support to do flash/reset/term on an IoT-LAB node.
It should also allow running test using 'testrunner'.

It handles both running from your computer or from the ssh frontend.

    makefile.iotlab.single: add IoT-LAB testbed support for one node
    
    Add support to do flash/reset/term on an IoT-LAB node.
    It also allow running test using 'testrunner'.
    
    Configuration variables are:
    
    * `IOTLAB_NODE` which should be set to your node url
      * The full url including site to use from your computer `m3-1.grenoble.iot-lab.info`
      * The short url when used on the IoT-LAB frontend `m3-1`
    * `IOTLAB_EXP_ID` for your experiment id for flash and reset.
      By default it tries to use your currently running experiment if you have only one
    * `IOTLAB_USER`: is read from `${HOME}/.iotlabrc` as saved by `iotlab-auth`
      * It is expected to have run `iotlab-auth` beforehand.

Usage is 

    make IOTLAB_NODE=m3-1.iot-lab.info BOARD=iotlab-m3 all flash term

Also, it is possible to detect automatically a first node to use from available ones:

    make IOTLAB_NODE=auto-ssh BOARD=iotlab-m3 all flash term

TODO:

* [x] Verify it's working on the frontend (not tested yet)
* [x] Run the release tests with it to find issues
* [x] Reword first commit message

### Issues/PRs references

Allow running 'make test' using IoT-LAB nodes.

### CODE REVIEW

**Comment by miri64** [comment]
Because I only notice it now:

**Comment by aabadie** [comment]
Maybe this PR could be generalized to the new other nodes supported on iotlab (samr21-xpro, b-l072z-lrwan1, microbit) ?

**Comment by aabadie** [comment]
Maybe this PR could be generalized to the new other nodes supported on iotlab (samr21-xpro, b-l072z-lrwan1, microbit, arduino-zero) ?

**Comment by cladmi** [comment]
I did a rewrite in this direction (rename FLASHFILE to IOTLAB_FLASHFILE).

From the code it should already work if included in the board and `IOTLAB_FLASHFILE` defined.
I could also overwrite the TERM/RESET instead of lazy setting them and choose to include in Makefile.include.
Not sure if the `ifneq (,$(IOTLAB_NODE))` protection should be in the file or when importing the file then.

Do they all flash with an ELFFILE (except wsn430 of course) ?

**Comment by aabadie** [comment]
> Do they all flash with an ELFFILE

for samr21-xpro, b-l072z-lrwan1, micro-bit and arduino-zero, yes. Internally the gateway code does the conversion to bin or hex when required (using objcopy).

**Comment by cladmi** [comment]
I now included the file globally and overwrite the TERM/FLASH/RESET variables so it works with every node.

I tested it from my computer with m3, a8, wsn430, arduino-zero, samr21. I am able to run the `make test`.
You can test it with the byplnfqfr-lrwan1 and give some feedback.

**Comment by aabadie** [comment]
nice !

> byplnfqfr-lrwan1

Feel free to complain against ST if you also don't like the marketing name of this board ;)
On iot-lab it's called st-lrwan1, which is a little shorter and simpler to type. Maybe we could use the same name in RIOT by default. What do you think ?
Also on iotlab, users need to be part of a specific group to have access to this kind of nodes (this restriction was chosen because of the long range radio usable on public radio bands)

**Comment by cladmi** [comment]
I think I will also add a `IOTLAB_NODE=auto` handling, using a board->iotlab-archi mapping to get the first matching node in the experiment.

**Comment by cladmi** [comment]
I added support for node detection:
 * IOTLAB_NODE=auto for on the ssh frontend
 * IOTLAB_NODE=auto-ssh for running on your computer through ssh.

It uses a internal mapping between RIOT BOARD and iot-lab archi.

It selects the first node with deployment == 0, matching architecture (and on the current frontend for auto).

Only tested from my computer at the moment too.

File: dist/riot-gitconfig [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[user]
```
Comment (aabadie) [inline dist/riot-gitconfig]: Is this file needed ?

Comment (cladmi) [reply dist/riot-gitconfig]: Which file ? I do not see any file there :p

**Comment by cladmi** [comment]
* [ ] Missing: extract the flash/reset return value and put error when it fails:
    * [ ] iotlab-node
    * [ ] iotlab-ssh

```
iotlab-node  --list euratech,wsn430,42 --update /home/harter/work/git/RIOT/tests/xtimer_remove/bin/wsn430-v1_4/tests_xtimer_remove.hex
{
    "1": [
        "wsn430-42.euratech.iot-lab.info"
    ]
}
```

**Comment by cladmi** [comment]
IoT-LAB commands now check the return value. It will need https://github.com/iot-lab/ssh-cli-tools/pull/36 to be working properly for A8 nodes.

**Comment by cladmi** [comment]
This is now in a ready state. I tested on the IoT-LAB frontend too.

File: dist/testbed-support/makefile.iotlab.single.inc.mk [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+#
+# If the user has multiple running experiments, the one to use must configured
+# with the `IOTLAB_EXP_ID` setting
+#
+# Prerequisites
+# -------------
+#
+# * Install iotlab-cli-tools: https://github.com/iot-lab/cli-tools
+# * Install iotlab-ssh-cli-tools: https://github.com/iot-lab/ssh-cli-tools
+#   * Can be done using given `requirements.txt`
```
Comment (aabadie) [inline dist/testbed-support/makefile.iotlab.single.inc.mk]: This is unclear

File: dist/testbed-support/makefile.iotlab.single.inc.mk [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+# * Configured your account with 'iotlab-auth'
```
Comment (aabadie) [inline dist/testbed-support/makefile.iotlab.single.inc.mk]: Configure

File: dist/testbed-support/makefile.iotlab.single.inc.mk [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+# * If using an A8 node from a frontend:
+#   * Register the frontend public key in your ssh keys list on your profile.
```
Comment (aabadie) [inline dist/testbed-support/makefile.iotlab.single.inc.mk]: Maybe give the tutorial link that explains this

File: dist/testbed-support/makefile.iotlab.single.inc.mk [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+#
+# Additional for A8 nodes
+#
+# * Install iotlab-ssh-cli-tools https://github.com/iot-lab/ssh-cli-tools
```
Comment (aabadie) [inline dist/testbed-support/makefile.iotlab.single.inc.mk]: This is already the second point of this list

File: dist/testbed-support/makefile.iotlab.single.inc.mk [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+ifeq (,$(IOTLAB_NODE))
+  $(warning IOTLAB_NODE undefined it should be defined to:)
```
Comment (aabadie) [inline dist/testbed-support/makefile.iotlab.single.inc.mk]: Maybe add a dot or comma after `undefined`

File: dist/testbed-support/makefile.iotlab.single.inc.mk [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+ifneq (,$(filter %.iot-lab.info, $(IOTLAB_NODE)))
+  _IOTLAB_ON_FRONTEND =
+else
+  _IOTLAB_ON_FRONTEND = 1
+endif
+# Work with node url without 'node-'
+override IOTLAB_NODE := $(patsubst node-%,%,$(IOTLAB_NODE))
+
+
+# Create node list and optionnaly frontend url
```
Comment (aabadie) [inline dist/testbed-support/makefile.iotlab.single.inc.mk]: optionally

File: dist/testbed-support/makefile.iotlab.single.inc.mk [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+ifeq (,$(_IOTLAB_ON_FRONTEND))
+  # m3-380.grenoble.iot-lab.info    -> grenoble,m3,380
+  # a8-1.grenoble.iot-lab.info      -> grenoble,a8,1
+  _NODELIST_SED := 's/\([^.]*\)-\([^.]*\).\([^.]*\).*/\3,\1,\2/'
```
Comment (aabadie) [inline dist/testbed-support/makefile.iotlab.single.inc.mk]: So it is possible to determine the BOARD from the IOTLAB_NODE. That would be a great feature!

File: dist/testbed-support/makefile.iotlab.single.inc.mk [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+#
+# It can be run:
+# * From your computer by setting IOTLAB_NODE to the full url like
+#   m3-380.grenoble.iot-lab.info or a8-1.grenoble.iot-lab.info
+# * From IoT-LAB frontend by setting IOTLAB_NODE to the short url like
+#   m3-380 or a8-1
+#
+# # Usage
+#
+#   make BOARD=iotlab-m3 IOTLAB_NODE=m3-380.grenoble.iot-lab.info flash term
```
[the remark this answers is not in the document]
Comment (cladmi) [reply dist/testbed-support/makefile.iotlab.single.inc.mk]: BOARD is used before in the build system, even in the application Makefile so would not want to touch this. It goes in the same problems as board aliases.

But I did the opposite and allowed `IOTLAB_NODE=auto-ssh make BOARD=iotlab-m3 flash term`

File: dist/testbed-support/makefile.iotlab.single.inc.mk [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+# Command to check if 'stdin' is 0. Cannot use 'cmp - <(echo 0)' without bash shell
+_STDIN_EQ_0 = grep 0
+
+ifneq (iotlab-a8-m3,$(BOARD))
+
+  # M3 and wsn430 nodes
+  FLASHER     = iotlab-node
+  RESET       = iotlab-node
+  _NODE_FMT   = --jmespath='keys(@)[0]' --format='int'
+  FFLAGS      = $(_NODE_FMT) $(_IOTLAB_EXP_ID) $(_IOTLAB_NODELIST) --update $(IOTLAB_FLASHFILE) | $(_STDIN_EQ_0)
```
Comment (aabadie) [inline dist/testbed-support/makefile.iotlab.single.inc.mk]: Maybe rename this variable FLASH_FLAGS ?

**Comment by aabadie** [comment]
> There's this potential improvement regarding BOARD variable automatic setup based on IOTLAB_NODE.

Ok, you answered before I fully submit my review.

File: dist/testbed-support/makefile.iotlab.single.inc.mk (hunk 2 above)
Comment (cladmi) [reply dist/testbed-support/makefile.iotlab.single.inc.mk]: Problem, do I explain with `sudo pip install -r` or the right way `pip install --user -r` and then expect them to setup their `PATH` properly ? That is why I linked with the official documentation for it.

It was not explained how to install the cli-tools in the other one either.

That makes me think I did not put information in the `README.iotlab`.

File: dist/testbed-support/makefile.iotlab.single.inc.mk (hunk 4 above)
Comment (cladmi) [reply dist/testbed-support/makefile.iotlab.single.inc.mk]: Which one is it ?

Also I expect people to know how to use IoT-LAB anyway for every other steps.

Note: it is only required because `iotlab-ssh` always uses an ssh gateway even when not required.

File: dist/testbed-support/makefile.iotlab.single.inc.mk (hunk 5 above)
Comment (cladmi) [reply dist/testbed-support/makefile.iotlab.single.inc.mk]: Oh I should remove it from the first list then, I thought I missed it.

File: dist/testbed-support/makefile.iotlab.single.inc.mk (hunk 2 above)
Comment (aabadie) [reply dist/testbed-support/makefile.iotlab.single.inc.mk]: Indeed, there's no installation section in the iotlabcli readme but there's one in the iotlabsshcli. Maybe you could just link to the related cli-tools and ssh-cli-tools tutorials on the IoT-LAB website ?

File: dist/testbed-support/makefile.iotlab.single.inc.mk (hunk 4 above)
Comment (cladmi) [reply dist/testbed-support/makefile.iotlab.single.inc.mk]: When doing it from your computer it also requires that you add your personal key to the website…

Comment (aabadie) [reply dist/testbed-support/makefile.iotlab.single.inc.mk]: This one: https://www.iot-lab.info/tutorials/ssh-cli-client/

File: dist/testbed-support/makefile.iotlab.single.inc.mk (hunk 8 above)
Comment (cladmi) [reply dist/testbed-support/makefile.iotlab.single.inc.mk]: Not entirely true, because of when BOARD should be set, and also `wsn430-1` can have two different types so would require fetching the API to know.

File: dist/testbed-support/makefile.iotlab.single.inc.mk (hunk 10 above)
Comment (cladmi) [reply dist/testbed-support/makefile.iotlab.single.inc.mk]: Do a `git grep FFLAGS` in RIOT. It is not mine.

File: dist/testbed-support/makefile.iotlab.single.inc.mk (hunk 6 above)
Comment (cladmi) [reply dist/testbed-support/makefile.iotlab.single.inc.mk]: In fact this message should never appear as it is already checked in Makefile.include, but I used it as a documentation.
Will update anyway.

File: dist/testbed-support/makefile.iotlab.single.inc.mk (hunk 4 above)
Comment (cladmi) [reply dist/testbed-support/makefile.iotlab.single.inc.mk]: I removed the mention on `requirements.txt` entirely.
My official documentation is use the IoT-LAB github documentation to install, I only added requirements.txt as version reference and for advanced users.

File: dist/testbed-support/makefile.iotlab.single.inc.mk (hunk 2 above)
Comment (cladmi) [reply dist/testbed-support/makefile.iotlab.single.inc.mk]: I removed the mention on requirements.txt entirely.
My official documentation is use the IoT-LAB github documentation to install, I only added requirements.txt as version reference and for advanced users.

**Comment by cladmi** [comment]
I rebased, but then added a doc for the ssh key tutorial so you can review the change.
I will re-rebase after.

File: dist/testbed-support/makefile.iotlab.single.inc.mk [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+# If the user has multiple running experiments, the one to use must be
+# configured with the `IOTLAB_EXP_ID` setting
+#
+# Prerequisites
+# -------------
+#
+# * Install iotlab-cli-tools: https://github.com/iot-lab/cli-tools
+# * Configure your account with 'iotlab-auth'
+# * Register your computer public key in your ssh keys list on your profile:
+#   https://www.iot-lab.info/tutorials/ssh-cli-client/
```
Comment (aabadie) [inline dist/testbed-support/makefile.iotlab.single.inc.mk]: I think this link should go with the use of "A8 node from a frontend section" below. The correct tutorial for setting the SSH access is here: https://www.iot-lab.info/tutorials/configure-your-ssh-access/

Comment (cladmi) [reply dist/testbed-support/makefile.iotlab.single.inc.mk]: Ahhhh yeah.

Comment (cladmi) [reply dist/testbed-support/makefile.iotlab.single.inc.mk]: I knew I would miss something when I did it :D

File: dist/testbed-support/makefile.iotlab.single.inc.mk [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+  $(warning  * <type>-<number> when run from iot-lab frontend)
+  $(warning    Example: m3-380 or a8-1)
+  $(warning  * 'auto' or 'auto-ssh' to try auto-detecting the node from your experiment
+  $(error)
+endif
+
+IOTLAB_AUTH ?= $(HOME)/.iotlabrc
+IOTLAB_USER ?= $(shell cut -f1 -d: $(IOTLAB_AUTH))
+
+# Optionnal experiment id. Required when having multiple experiments
```
Comment (aabadie) [inline dist/testbed-support/makefile.iotlab.single.inc.mk]: typo Optional

File: dist/testbed-support/makefile.iotlab.single.inc.mk [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+# * From your computer by setting IOTLAB_NODE to the full url like
+#   m3-380.grenoble.iot-lab.info or a8-1.grenoble.iot-lab.info
+# * From IoT-LAB frontend by setting IOTLAB_NODE to the short url like
+#   m3-380 or a8-1
+#
+# # Usage
+#
+#   make BOARD=iotlab-m3 IOTLAB_NODE=m3-380.grenoble.iot-lab.info flash term
+#
+# It is the user responsability to start an experiment beforehand
```
Comment (aabadie) [inline dist/testbed-support/makefile.iotlab.single.inc.mk]: typo respons**i**bility

**Comment by cladmi** [comment]
I should really run a spellchecker on my documentation ><

**Comment by aabadie** [comment]
Let's merge this very useful PR
